### PR TITLE
Add support for nodeConfig in ApigeeEnvironment

### PR DESCRIPTION
**Pull request by xuchenma**
This PR solves https://github.com/hashicorp/terraform-provider-google/issues/12218

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
Add support for nodeConfig in ApigeeEnvironment
```

### CODE REVIEW

**Comment by xuchenma** [comment]
Hi Riley, this API isn't public yet (will be there later this week) so the tests won't pass for now.
Meanwhile I think this PR supports the POST and GET calls, but not the PATCH call. Could you shed some light on that?
The PATCH call is
```
curl -s -H "$AUTH" "https://apigee.googleapis.com/v1/organizations/$ORG_ID/environments/$ENVIRONMENT_NAME?updateMask=node_config" \
  -X PATCH \
  -H "Content-Type:application/json" \
  -d '{                                                           
     "nodeConfig": {
        "minNodeCount":"'"$MIN_NODE_COUNT"'",
        "minNodeCount":"'"$MAX_NODE_COUNT"'",
     }
  }'
```
Thanks!

**Comment by rileykarson** [comment]
Sorry, lost email on my side. I'm OOO for a bit, so picked you a new reviewer from the lottery.

**Comment by c2thorn** [comment]
> Hi Riley, this API isn't public yet (will be there later this week) so the tests won't pass for now. Meanwhile I think this PR supports the POST and GET calls, but not the PATCH call. Could you shed some light on that? The PATCH call is
> 
> ```
> curl -s -H "$AUTH" "https://apigee.googleapis.com/v1/organizations/[](https://apigee.googleapis.com/v1/organizations/)$ORG_ID/environments/$ENVIRONMENT_NAME?updateMask=node_config" \
>   -X PATCH \
>   -H "Content-Type:application/json" \
>   -d '{                                                           
>      "nodeConfig": {
>         "minNodeCount":"'"$MIN_NODE_COUNT"'",
>         "minNodeCount":"'"$MAX_NODE_COUNT"'",
>      }
>   }'
> ```
> 
> Thanks!

It seems that this is the first updatable field introduced to the resource. We can specify `PATCH` via
`update_verb: :PATCH`.  see https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/apigee/api.yaml#L329

File: mmv1/products/apigee/api.yaml [hunk 1]
```diff
@@ -316,6 +316,31 @@ objects:
...
+        description: |
+          NodeConfig for setting the min/max number of nodes associated with the environment.
+        properties:
+          - !ruby/object:Api::Type::String
+            name: 'minNodeCount'
+            description: |
+              The minimum total number of gateway nodes that the is reserved for all instances that
+              has the specified environment. If not specified, the default is determined by the
+              recommended minimum number of nodes for that gateway.
+            input: true
```
Comment (c2thorn) [inline mmv1/products/apigee/api.yaml:330]: If this is updatable, you do not want `input: true`, as this makes the field ForceNew

**Comment by xuchenma** [comment]
> > Hi Riley, this API isn't public yet (will be there later this week) so the tests won't pass for now. Meanwhile I think this PR supports the POST and GET calls, but not the PATCH call. Could you shed some light on that? The PATCH call is
> > ```
> > curl -s -H "$AUTH" "https://apigee.googleapis.com/v1/organizations/[](https://apigee.googleapis.com/v1/organizations/)$ORG_ID/environments/$ENVIRONMENT_NAME?updateMask=node_config" \
> >   -X PATCH \
> >   -H "Content-Type:application/json" \
> >   -d '{                                                           
> >      "nodeConfig": {
> >         "minNodeCount":"'"$MIN_NODE_COUNT"'",
> >         "minNodeCount":"'"$MAX_NODE_COUNT"'",
> >      }
> >   }'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Thanks!
> 
> It seems that this is the first updatable field introduced to the resource. We can specify `PATCH` via `update_verb: :PATCH`. see https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/apigee/api.yaml#L329

The additional complication is that, actually there is already a PUT call that can be used to update all the other fields (except for this nodeConfig) of this resource (ref: https://cloud.google.com/apigee/docs/reference/apis/apigee/rest/v1/organizations.environments/update), and this new "nodeConfig" can only be updated via PATCH.
If PUT is used to updated "nodeConfig", or PATCH is used to update any field other than "nodeConfig", the call will fail.
I'm not sure how this situation can be represented in magic-modules..

**Comment by c2thorn** [comment]
> The additional complication is that, actually there is already a PUT call that can be used to update all the other fields (except for this nodeConfig) of this resource (ref: [cloud.google.com/apigee/docs/reference/apis/apigee/rest/v1/organizations.environments/update](https://cloud.google.com/apigee/docs/reference/apis/apigee/rest/v1/organizations.environments/update)), and this new "nodeConfig" can only be updated via PATCH. If PUT is used to updated "nodeConfig", or PATCH is used to update any field other than "nodeConfig", the call will fail. I'm not sure how this situation can be represented in magic-modules..

I see. I believe you can set the `update_verb` property on the field itself, and we have logic to move it to a separate call

**Comment by xuchenma** [comment]
Thanks, I added `update_verb` and `update_url` to the field, and generated a local provider.
It seems that a code block that starts with `if d.HasChange("node_config") {` is generated here https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_apigee_environment.go#L300
However, a PUT call was fired prior to this, at line https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_apigee_environment.go#L285.

When I tested it locally and updated the nodeConfig, it failed with a PUT call failure. But when I commented out the PUT call (above line 285 and a few related lines), the PATCH was fired properly and the update succeeded.

Is there a way to not fire the PUT call when only the nodeConfig has changed?

**Comment by c2thorn** [comment]
> Thanks, I added `update_verb` and `update_url` to the field, and generated a local provider. It seems that a code block that starts with `if d.HasChange("node_config") {` is generated here [https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_apigee_environment.go#L300](https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_apigee_environment.go?rgh-link-date=2022-08-17T00%3A25%3A38Z#L300) However, a PUT call was fired prior to this, at line [https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_apigee_environment.go#L285](https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_apigee_environment.go?rgh-link-date=2022-08-17T00%3A25%3A38Z#L285).
> 
> When I tested it locally and updated the nodeConfig, it failed with a PUT call failure. But when I commented out the PUT call (above line 285 and a few related lines), the PATCH was fired properly and the update succeeded.
> 
> Is there a way to not fire the PUT call when only the nodeConfig has changed?

Apologies, that was not expected. I'll have to take a look myself to see how to best handle this situation as it's fairly rare. Would you mind pushing the changes you had in MM locally that ended up failing? So I can pull the change and test things out on my end?

**Comment by xuchenma** [comment]
Sure, pushed, thanks!

**Comment by c2thorn** [comment]
@xuchenma 
I took a look at this. What I didn't realize is that any PUT body needs to have the `nodeConfig` field present because the lack of `nodeConfig` apparently implies you are removing it (which is an update). Which means it's required in the PUT body but not allowed to be updated.

This is an API constraint that I don't believe we've run into before for MMv1 generated resources. My initial impression is that this should be fixed in the API itself. I imagine TF is likely one of the top consumers of the API, and this constraint is not helpful. Surely this must be an implementation issue and not an intended design decision? I don't know if you have any insight into that.

Attempting to work around this in MMv1 generation will be relatively unsatisfactory. How we have the PR now excludes `nodeConfig` from the `PUT` request. There is no way currently in MMv1 to not do some sort of request in the update method. The only way I can think to workaround is to add `update_encoder` logic that adds the old `nodeConfig` value to the `PUT` request, something like (untested)
```
old, _ = d.GetChange("node_config")
obj["nodeConfig"] = old
```
This should allow the `PUT` to succeed leading to the actual `PATCH` request after, but then we are running an unnecessary non-op `PUT` request when only `node_config` has changed. This should be a relatively uncommon situation, but is still unsatisfactory... let me know what you think.

The only non-compromising way to implement this would be to convert the resource to be completely handwritten with very specific custom logic to deal with this.

**Comment by c2thorn** [comment]
example an `update_encoder`

https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/templates/terraform/update_encoder/pubsub_topic.erb

reference to the file in `terraform.yaml`: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/pubsub/terraform.yaml#L74

**Comment by c2thorn** [comment]
I believe the logic within the update encoder file will look like
```
old, _ = d.GetChange("node_config")
obj["nodeConfig"] = old
return obj, nil
```

**Comment by xuchenma** [comment]
Thanks Cameron! I made the changes and it worked for me locally. Please check.

There are a few complications:
1. The PUT call is not an async call. The current implementation fails when any meaningful PUT is called, because it tries to check the "async" operation status based on operation_id, but the PUT call doesn't return one. This is a separate bug but I fixed it all together here, by checking async only for "create" and "delete" on ApigeeEnvironment. It would be better if we can add async status check also for the PATCH only.
2. The call `resourceApigeeEnvironmentUpdateEncoder` is generated before both PUT and PATCH. It should only update obj["nodeConfig"] for PUT, so I put a check at the top.
3. There are some type conversions that involve []interface{} and map[string]interface{}.
4. `current_aggregate_node_count` is output-only and need to be removed, before obj["nodeConfig"] can be served as input.

**Review by c2thorn — CHANGES_REQUESTED** [review]
Looks good so far, but unfortunately I'm only just now noticing that we do not have explicit tests for update. We need to add a handwritten test that updates `node_config` to ensure that it works now and will continue working in the future.

https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/resources/resource_cloudfunction2_function_test.go.erb should be a very similar example, with further documentation at https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform#update-tests

**Comment by xuchenma** [comment]
Thanks for the review! I added a test for nodeConfig update; in my local I can see the test is calling PATCH correctly. However as I mentioned in the above comment https://github.com/GoogleCloudPlatform/magic-modules/pull/6349#issuecomment-1220080526, PUT is a sync call, but PATCH is an async call. Because I removed the async check for update (otherwise PUT is broken), in the test, after calling PATCH, it immediately calls GET to verify the state, while the state has not changed yet.
Is there a way to add async check for the PATCH only?

**Comment by c2thorn** [comment]
@xuchenma I see, sorry for not catching that in your earlier message. we are really trying to stretch the MMv1 template with this logic here, and I am getting close to just recommending you to copy the generated files and convert the resource to handwritten so you can tweak the go code specifically.

However, there is another attempt I can think of. This is the resource template where we write the logic for fields with custom updates: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/templates/terraform/resource.erb#L702
You'll notice there is no logic to try to do asynchronous calls. To try to shoehorn this logic into the template (without changing the template, as that could cause more disruption than just switching to handwritten), I think we can change the resource's update verb to `PATCH` and flip every other field to use a custom `PUT` verb. That way, `PATCH` gets the async logic, and `PUT` is handled separately. What do you think?

**Comment by c2thorn** [comment]
Hi @xuchenma, I have spoken with Riley and he is also open to moving this to a handwritten resource if needed. To be frank, I cannot be 100% confident that my previous comment would also not produce some sort of complication as I have been wrong up to this point. 

I apologize for the amount of iterations this has caused you, but this really is behavior we have not run into otherwise. It's up to you if you want to explore my earlier suggestion, but know that converting to handwritten will give you full control of the go code and let you tweak the update function to exactly what is needed here. 

You would not be starting from scratch as you could re-use the majority of the generated code as shown in https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6349-old..auto-pr-6349 . There's some extra steps like manually registering the resource into https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/utils/provider.go.erb#L377

**Comment by xuchenma** [comment]
Thanks Cameron and Riley. I tried to change the update_verb of ApigeeEnvironment resource to PATCH just now, it seems to be working for nodeConfig. However, setting update_verb to PUT for other fields doesn't seem to generate another PUT codeblock in resource_apigee_environment.go, so there could be more complications.

But as I look deeper, I think with our current implementation for ApigeeEnvironment in mmv1, we can't really make a valid update call. All the fields except for "name" is annotated with "input: true" which results in ForceNew when its value changes, while making changes to "name" is not allowed in the API backend (we should add "input: true" to "name" too). So, effectively no update call can be made to ApigeeEnvironment. Although the API supports updating some of the fields like "properties", such field is not added to mmv1 yet.

I think we should go for handwritten resource when we need to support PUT updates to ApigeeEnvironment for new fields, but right now we can just set the update_verb of the resource as PATCH since it is not a breaking change. What do you think?

**Comment by c2thorn** [comment]
> 
> I think we should go for handwritten resource when we need to support PUT updates to ApigeeEnvironment for new fields, but right now we can just set the update_verb of the resource as PATCH since it is not a breaking change. What do you think?

That sounds reasonable for now. There's a chance the `PUT` update_verb will actually work correctly when we add an updatable field as well.

**Comment by xuchenma** [comment]
Thanks, updated!

**Comment by c2thorn** [comment]
Running Apigee tests in test build since they don't run in VCR

**Review by c2thorn — APPROVED** [review]
All tests passed including the new update. Generated code looks good to me. Thanks for your efforts on this @xuchenma !

**Pull request merged**